### ISSUE TITLE
Wrong type hint for asyncio.base_events.Server sockets attribute

### ISSUE THREAD

**Issue by ramalho** [python/typeshed#5535]
The fix for issue #2364 ("AbstractServer" has no attribute "sockets") added this declaration:

```python
class AbstractServer:
    sockets: Optional[List[socket]]
```

However, in the Python 3.9 [implementation](https://github.com/python/cpython/blob/3.9/Lib/asyncio/base_events.py#L324) of `asyncio.base_events.Server`, `sockets` is a property defined like this:

```python
    @property
    def sockets(self):
        if self._sockets is None:
             return ()
        return tuple(trsock.TransportSocket(s) for s in self._sockets)
```

So we have two problems:

1) The value of `sockets` is never `None`, it's always a `tuple`. 
2) It is really a `tuple[TransportSocket, ...]`.

There is [example code](https://docs.python.org/3/library/asyncio-stream.html#tcp-echo-server-using-streams) in the `asyncio` docs and [test code](https://github.com/python/cpython/blob/b798ab06937f8bb24b444a49dd42e11fff15e654/Lib/test/test_asyncio/test_server.py#L55) in the standard library that assume `sockets` is a sequence of sockets.

@ymyzk contributed the patch for #2364 noting "I noticed that this class has some changes in Python 3.7, so we need to update it again in the near future."

The documentation for [`asyncio.Server.sockets`](https://docs.python.org/3.7/library/asyncio-eventloop.html#asyncio.Server.sockets) states:

> Changed in version 3.7: Prior to Python 3.7 Server.sockets used to return an internal list of server sockets directly. In 3.7 a copy of that list is returned.

As I mentioned, the actual value returned is a tuple of zero or more `asyncio.Server.TransportSocket` objects.

**Comment by ramalho** [comment]
The documentation for [`asyncio.Server.sockets`](https://docs.python.org/3.7/library/asyncio-eventloop.html#asyncio.Server.sockets) states:

> Changed in version 3.7: Prior to Python 3.7 Server.sockets used to return an internal list of server sockets directly. In 3.7 a copy of that list is returned.

As I mentioned before, the actual value returned is a tuple of zero or more `asyncio.Server.TransportSocket` objects.

I added type hints to the `asyncio` example titled [_TCP echo server using streams_](https://docs.python.org/3/library/asyncio-stream.html#tcp-echo-server-using-streams) from the official documentation. Mypy wrongly complains that `Value of type "Optional[List[socket]]" is not indexable` pointing at this line:

```python
    addr = server.sockets[0].getsockname()
```

Code reading `sockets[0]` [also appears](https://github.com/python/cpython/blob/b798ab06937f8bb24b444a49dd42e11fff15e654/Lib/test/test_asyncio/test_server.py#L55) in the `asyncio` test suite.

**Comment by srittau** [comment]
Quick check:

* 3.6.10: simple attribute, can be `None`
* 3.7.10: property, returns `list`, never `None`
* 3.8.3: property, returns `tuple`, never `None`

**Comment by srittau** [comment]
Also, `socket` is defined on `AbstractServer` in typeshed, not on `Server`. On all checked versions, `AbstractServer` doesn't defined `socket` at all.

**Comment by ramalho** [comment]
It's worth noting that [`asyncio.trsock.TransportSocket`](https://github.com/python/cpython/blob/main/Lib/asyncio/trsock.py#L5) is explicitly **not** a subtype of `socket.socket`. Its docstring states:

> All potentially disruptive operations (like "socket.close()") are banned.

**Comment by ramalho** [comment]
Thank you @srittau fro addressing this so quickly!

I saw you annotated the attribute as `Iterable`. 

However, both examples that I mentioned from the standard library use it as a `Sequence` like this:

```python
    addr = server.sockets[0].getsockname()
```

The examples are:

- [example code](https://docs.python.org/3/library/asyncio-stream.html#tcp-echo-server-using-streams) in the `asyncio` docs;
- test code in [Lib/test/test_asyncio/test_server.py](https://github.com/python/cpython/blob/b798ab06937f8bb24b444a49dd42e11fff15e654/Lib/test/test_asyncio/test_server.py#L55).

**Comment by srittau** [comment]
@ramalho The attributes/properties are annotated as `list` or `Tuple`. Only the argument on 3.7+ is annotated as `Iterable` and that should work, as the implementation uses `list(self._sockets)` or `tuple(self._sockets)`.

**Comment by ramalho** [comment]
I am sorry but don't understand your answer, @srittau. Yes, the annotation is compatible with the implementations, but the `Iterable` type is too general.

The documentation for [`asyncio.Server.sockets`](https://docs.python.org/3.7/library/asyncio-eventloop.html#asyncio.Server.sockets) in Python 3.7 and later states that the attribute returns a `list`.

> Changed in version 3.7: Prior to Python 3.7 Server.sockets used to return an internal list of server sockets directly. In 3.7 a copy of that list is returned. 

Users expect to be able to index a `list`, and both examples from CPython itself (docs and tests) do this:

```python
    addr = server.sockets[0].getsockname()
```

Therefore, the `Iterable` annotation will break user's code that uses `sockets` as a `Sequence`, according to the Python ≥ 3.7 documentation and examples provided in the distribution.

**Comment by srittau** [comment]
But the `sockets` attribute isn't annotated with `Iterable`. It's annotated with (depending on the version) `Optional[list[_socket]]`, `list[_socket]`, or `Tuple[_socket, ...]`.

**Comment by ramalho** [comment]
OK, thanks! Sorry for the misunderstanding.

**Comment by layday** [comment]
FYI, the [code sample](https://docs.python.org/3/library/asyncio-stream.html#tcp-echo-server-using-streams) from the asyncio docs still does not type check because `asyncio.create_server` returns an instance of `AbstractServer` which does not implement `sockets`, and you can't even use an assertion prior to accessing `sockets` anymore.

**Comment by JelleZijlstra** [comment]
`assert isinstance(server, Server)` should work, right?

Also, should we make create_server() return a Server instead of an AbstractServer in the stub?

**Comment by layday** [comment]
> `assert isinstance(server, Server)` should work, right?

That works but it's not exposed in `__init__` and would have to be imported from `base_events`.

> Also, should we make create_server() return a Server instead of an AbstractServer in the stub?

That'd fix it for the top-level `start_server` function. `asyncio.get_{event,running}_loop()` would have to return a `BaseEventLoop` for the loop's `create_server` method to return a `Server` as well.

**Comment by srittau** [comment]
Cf. #5566. `loop.create_server()` is documented to return `Server`, not `AbstractServer` anyway.

**Comment by layday** [comment]
Great, thanks!
